CallsAndSMS/app/src/main/java/nisrulz/github/example/callsandsms/MainActivity.kt

```diff
@@ -5,157 +5,134 @@ import android.content.Intent
-                            .show()
+                        showToast("Permission not granted yet!")
                     }
                 } else {
-                    Toast.makeText(this@MainActivity, "Enter a phone number", Toast.LENGTH_SHORT)
-                        .show()
+                    showToast("Enter a phone number")
                 }
             }
             btnSendMessage.setOnClickListener {
                 val message = etMessage.text.toString()
                 val phoneNo = etPhoneNo.text.toString()
                 if (!TextUtils.isEmpty(message) && !TextUtils.isEmpty(phoneNo)) {
-                    val smsIntent = Intent(Intent.ACTION_SENDTO, Uri.parse("smsto:$phoneNo"))
+                    val smsIntent = Intent(
+                        Intent.ACTION_SENDTO, Uri.parse(
+                            createUriForProtocol(
+                                SMS_PROTOCOL, phoneNo
+                            )
+                        )
+                    )
                     smsIntent.putExtra("sms_body", message)
                     startActivity(smsIntent)
                 }
             }
-            btnSendMessageDirectly.setOnClickListener {
-                val message = etMessage.text.toString()
-                val phoneNo = etPhoneNo.text.toString()
-                if (!TextUtils.isEmpty(message) && !TextUtils.isEmpty(phoneNo)) {
-                    // Requires Permission to be declared in manifest
-                    // <uses-permission android:name="android.permission.SEND_SMS"/>
-                    // Then check and request for permission during runtime
-                    if (isPermissionGranted(Manifest.permission.SEND_SMS)) {
-                        val smsManager = SmsManager.getDefault()
-                        smsManager.sendTextMessage(phoneNo, null, message, null, null)
-                    } else {
-                        Toast.makeText(this@MainActivity, "Permission denied", Toast.LENGTH_SHORT)
-                            .show()
-                    }
-                }
-            }
         }
     }
 
-    override fun onResume() {
-        super.onResume()
-        telephonyManager.listen(phoneStateListener, PhoneStateListener.LISTEN_CALL_STATE)
+    private fun createUriForProtocol(protocol: String, data: String) = "$protocol:$data"
+
+    private val requestPermissionLauncher =
+        registerForActivityResult(
+            ActivityResultContracts.RequestMultiplePermissions()
+        ) { }
+
+    private fun checkForPermissions() {
+        if (!allPermissionsGranted()) {
+            askForPermission()
+        }
+    }
+
+    /**
+     * Checks if all the permissions mentioned in [REQUIRED_PERMISSIONS]
+     * are granted or not.
+     */
+    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all { permission ->
+        ContextCompat.checkSelfPermission(
+            this, permission
+        ) == PackageManager.PERMISSION_GRANTED
     }
 
```

Comment:
The library sounds like a good option. But, considering this repository as an individual example of different scenarios so that anyone can focus on pure android stuff rather. So, I am not inclined towards using the library.